pkg/resources/manager/manager.go

```diff
@@ -257,18 +281,26 @@ func (r *resourceManager) allocate(machine *api.Machine, requiredResources core.
 	}
 
 	totalAllocatedRes := core.ResourceList{}
-	for _, s := range r.sources {
-		allocatedRes := s.Allocate(requiredResources)
+	var allocatedRes core.ResourceList
+	for key := range requiredResources {
+		s, ok := r.registredResources[key]
+		if !ok {
+			return fmt.Errorf("failed to find source for resource %s: %w", key, ErrResourceUnsupported)
+		}
 
-		sourceResourceList := s.GetAvailableResource()
-		sourceResourceQuantity := sourceResourceList[core.ResourceName(s.GetName())]
-		if sourceResourceQuantity.Sign() == -1 {
-			_ = s.Deallocate(requiredResources)
-			return ErrResourceNotAvailable
+		allocatedRes, err = s.Allocate(requiredResources)
+		if err != nil {
+			break
 		}
+
 		mergeResourceLists(totalAllocatedRes, allocatedRes)
 	}
 
+	if err != nil {
+		r.deallocateUnassignResources(totalAllocatedRes)
```

Comment:
Can't it be moved to line 293, in place of break ?

Reply:
fixed